docs: add AI Agent Runtime capability page (Agent Sessions API) (#2345)
pomerium/documentation: content/docs/capabilities/agentic-runtime.mdx

---
# cSpell:ignore runid Codex OpenCode kubelet serviceaccount autonumber refreshable Kata uncredentialed misroute

title: 'Run AI Agents in Isolated Sandboxes'
sidebar_label: 'AI Agent Runtime'
lang: en-US
description: 'Run any agent harness in a sandbox that holds no secrets, with Pomerium binding every request to a human approval and an attested pod.'
keywords:
  [
    AI agents,
    agent sandbox,
    agentic runtime,
    agent sessions API,
    run identity,
    workload identity,
    secret injection,
    egress control,
    prompt injection,
    MCP,
    LLM security,
    Claude Code,
  ]
---

# Run AI Agents in Isolated Sandboxes

You want an agent to do real work: read tickets, query a database, call an internal API, open a pull request. Real work needs credentials. But an agent is a program that takes instructions from untrusted text. Give it an API key, and you have given the key to anyone who can influence its input — a hostile issue comment, a poisoned dependency README, a malicious tool description.

> The usual mitigations try to make the agent careful. Careful is not a security boundary. The alternative is an architecture in which the agent has nothing to leak: it runs in a sandbox with no secrets, and everything it is allowed to reach sits behind a gateway that decides, per request and deterministically, whether this agent, acting for this person, may reach this resource.

Reviewer: maybe mention an example of this like hooks in a harness?

Pomerium — an open-source identity-aware proxy — provides that architecture as one stack. You choose the harness (Claude Code, Codex, Gemini CLI, OpenCode, or your own program) and the model; Pomerium provides the sandbox lifecycle, run identity, policy enforcement, and egress control. A Slack bot ships with it as a reference client, built on the same public API as anything you build.

:::info Availability

The agentic runtime is in active development and available for early access. [Contact us](https://www.pomerium.com/contact-sales) to get started, or see [MCP support](/docs/capabilities/mcp) for the generally available building blocks it is built on.

:::

## Architecture

<span className="themedDiagram themedDiagram--light">

![Agentic runtime architecture: clients, the Agent Sessions API, sandbox pods, the Pomerium gateway, and upstream services](./img/agentic-runtime/architecture.svg)

</span><span className="themedDiagram themedDiagram--dark">

![Agentic runtime architecture: clients, the Agent Sessions API, sandbox pods, the Pomerium gateway, and upstream services](./img/agentic-runtime/architecture-dark.svg)

</span>

The layers, from enforcement up to product:

> - **Pomerium** is the gateway. It authenticates every request against your identity provider, evaluates a deterministic policy, injects the credentials the upstream needs, and logs the result. In the reference configuration, every credentialed request a sandbox makes leaves through it.

Reviewer: This sounds a bit weird the last phrase in this bullet point.
```suggestion
- **Pomerium** is the gateway. It authenticates every request against your identity provider, evaluates a deterministic policy, injects the credentials the upstream needs, and logs the result. In the reference configuration, all outbound requests from a sandbox that require credentials are routed through Pomerium.
```

- **The sandbox pod** runs your agent next to a sidecar. The agent container holds no credentials of any kind. The sidecar holds the pod's credentials and proxies the agent's traffic to Pomerium.

> - **The Agent Sessions API** manages sessions: it claims sandbox pods, launches agents in them, streams their output as durable events, and suspends and revives them.

Reviewer: Probably need to mention destroying the pods too?

```suggestion
- **The Agent Sessions API** manages sessions: it claims sandbox pods, launches agents in them, streams their output as durable events, and suspends, revives them or terminates them.
```

The flow, end to end:

```mermaid
sequenceDiagram
    autonumber
    participant M as Session manager
    participant P as Pomerium
    participant A as Alice
    participant S as Sandbox sidecar
    M->>P: POST /agentic/runs (sealed pod identity, expected approver)
    P-->>M: run_id, approval_url
    S->>P: POST /agentic/token (projected token)
    P-->>S: 400 authorization_pending
    A->>P: opens approval_url, signs in, approves
    S->>P: POST /agentic/token (projected token)
    P-->>S: 200 run token (pom_art_…)
    S->>P: request to an upstream route (run token)
    P-->>S: policy evaluated, credentials injected
```

> Each hop carries a different credential: the manager's workload token, the person's SSO session, the pod's projected token, the run token. None of them can stand in for another.

Reviewer: Should manager be orchestrator to keep wording uniform? Or is manager something else in this context?

```suggestion
Each hop carries a different credential: the orchestrator's workload token, the person's SSO session, the pod's projected token, the run token. None of them can stand in for another.
```
